### ISSUE TITLE
[Automation API, multi-language components] Provider config statically set when package is imported for node, python, dotnet

### ISSUE THREAD

**Issue by EvanBoyle** [pulumi/pulumi#5582]
TLDR; provider config for node.js, python, and c# uses global state and is incompatible with Automation API inline programs. 

Provider config is instantiated statically at package import time:

```
import * as aws from "@pulumi/aws";

// any values here are static from the time the package was imported.
aws.config; 
```

See the implementation: https://github.com/pulumi/pulumi-aws/blob/master/sdk/nodejs/config/vars.ts

For traditional CLI driven pulumi programs this is fine as the order of ops is typically:

1. `pulumi config set aws:region us-west-2` from the CLI
2. `pulumi up` use CLI to exec a typescript program, this serialized config into environment variables
3. Pulumi up starts a langhost that execs the program with the appropriate config.
4. Program runs and `import * as aws from "@pulumi/aws"` establishes static config that cannot change.
5. Calls to `aws.config.*` have up to date config values.

With inline program the flow is problematic since we run a single program to set config and run the program:

1. We hit the import section of the program. `import * as aws from "@pulumi/aws"` is the first thing that runs, solidifying all values in `aws.config.*` (ie anything in the `aws:` config namespace is undefined).
2. A workspace and stack are created.
3. Config is set on the stack: `stack.setConfig("aws:region", { value: "us-west-2" });`
4. Program is run (`stack.up()`) and that config is serialized and handed off to the program.
5. The program gets the correct config via environment variables, but any provider config like that read from the awsx package here will return `undefined` as it was statically read from the import before we did any config setting: https://github.com/pulumi/pulumi-awsx/blob/b515c09c32196286380dc08b362c53272b63d6a3/nodejs/awsx/utils.ts#L137-L140

Our go SDK does the right thing here are exposes functions. This was the config is read fresh when the value in question is referenced: https://github.com/pulumi/pulumi-aws/blob/master/sdk/go/aws/config/config.go#L13

For node it appears we can use property getters to fix this in a non-breaking way.
Python definitely has the same behavior: https://github.com/pulumi/pulumi-aws/blob/master/sdk/python/pulumi_aws/config/vars.py
I'm not familiar with the lifecycle of C# static classes, but it looks like it's probably an issue here as well: https://github.com/pulumi/pulumi-aws/blob/master/sdk/dotnet/Config/Config.cs

### Languages
- [x] Dotnet (#7549)
- [x] Node.js (#7530)
- [x] Python (#7447)

**Comment by EvanBoyle** [comment]
For anyone reading, one workaround is to use environment variables where possible. For instance using `AWS_REGION`: https://github.com/pulumi/pulumi-aws/blob/master/sdk/nodejs/config/vars.ts#L41

**Comment by lukehoban** [comment]
A related issue (which I swore we had an open tracking issue on, but can't find now), is that `aws.config` doesn't work for first-class providers either.  `aws.config` always exposes the default provider configuration.  For that reason as well - I had assumed we would eventually need to make these config accessors functions (which can be passed providers, or are accessed through provider instances).  Sounds like this is another reason we will need to go that direction.

In practice, it is somewhat uncommon to need to access `aws.config` - and in most cases better to use a data source function.  But we definitely need to rethink how configuration values are exposed off of providers to handle both first-class providers and automation API.

**Comment by gitfool** [comment]
AFAICT, the static config properties in [sdk/dotnet/Config/Config.cs](https://github.com/pulumi/pulumi-aws/blob/master/sdk/dotnet/Config/Config.cs) are not accessed elsewhere in this project. Would it be a suitable workaround to simply defer initialization until first access by making the private field initialization lazy using [Lazy\<Pulumi.Config\>](https://docs.microsoft.com/en-us/dotnet/api/system.lazy-1)?

**Comment by tmccombs** [comment]
> one workaround is to use environment variables where possible

unfortunately that doesn't work for me, because my backend requires different credentials and may be in a different region than the provider for my stack. Also, as far as I know, there isn't a way to set up role assumption using environment variables.

**Comment by EvanBoyle** [comment]
@tmccombs you can set environment variables that are scoped to the workspace for automation api by passing in workspace options on creation. Not sure what language you are using, but here is an example for Go:

https://github.com/pulumi/automation-api-examples/blob/fb4ffddd8dd5eade5a7d1454e23123ae432fa3ac/go/inline_secrets_provider/main.go#L104-L111

**Comment by tmccombs** [comment]
@EvanBoyle maybe i'm missing something but wouldn't that environment variable be used for both the stack and the backend? It also doesn't solve the problem that there isn't an AWS_ASSUME_ROLE environment variable.

**Comment by komalali** [comment]
I spent some time today experimenting with how we might be able to do this in a non-breaking way for python. My hope was to turn `config` into a class and use `@property` getters to invoke the `__config__.get()` functions rather than have them statically declared. In any case, my experiments did not work out as I had hoped. I am unconfident that we can do this in a non-breaking way for python. 

Unless someone has an alternative idea, I would suggest introducing new functions to access the config values (like what we do in go) and deprecating the existing variables to be removed in 4.0

**Comment by justinvp** [comment]
For Python, I wonder if now that https://github.com/pulumi/pulumi/pull/6827 has landed if this is less of an issue for simple scenarios. The `config` module is now lazily loaded, so theoretically (I haven't had a chance to try it), if you only import `import pulumi_aws as aws` at the top of the file, and don't refer to `aws.config` anywhere, I would expect the config values won't be loaded until you actually refer to `aws.config`, which you could do inside your inline program.

Though, even if this does work, it wouldn't fully resolve the issue if you're doing something more complicated like using other code that happens to pull in `aws.config` before your inline program runs, or resetting the config values between multiple inline program runs, etc.

> My hope was to turn `config` into a class and use `@property` getters to invoke the `__config__.get()` functions rather than have them statically declared

Was your class a subclass of `types.ModuleType`? [PEP 549](https://www.python.org/dev/peps/pep-0549/) mentions the following:

```python
import sys, types
class _MyModuleType(types.ModuleType):
    @property
    def prop(self, instance, owner):
        ...

sys.modules[__name__].__class__ = _MyModuleType
```

I wonder if that'd work for us?

If not, [PEP 562](https://www.python.org/dev/peps/pep-0562/) is now available in Python 3.7, which allows using `__getattr__` in a module. I wonder if we could use that on 3.7+, and fallback to the old way for 3.6 (and possibly also when `TYPE_CHECKING`, if it'd help type checkers).

**Comment by komalali** [comment]
>The config module is now lazily loaded, so theoretically (I haven't had a chance to try it), if you only import pulumi_aws as aws at the top of the file, and don't refer to aws.config anywhere, I would expect the config values won't be loaded until you actually refer to aws.config, which you could do inside your inline program.

I suspect you might be correct for that narrow case, but yeah, that is a lot of caveats and it doesn't solve for the complex cases as you mentioned. I don't think PEP-549 will work as it was rejected in favor of PEP-562. 

I don't love the idea of falling back to the old way for 3.6 as the old way just plain doesn't work. If we want to continue supporting 3.6 it shouldn't be an afterthought. As a compromise, I would want to introduce new functions to the config module that would actually work for 3.6.

Anyway, I'll try out PEP 562 and report back.

**Comment by justinvp** [comment]
PEP 549 is about adding properties to modules, which was rejected in favor of PEP 562. However, the code snippet I showed from PEP 549 apparently already "works, and is supported behavior" according to the document. It just isn't as elegant as what the PEP was proposing of allowing properties directly on modules.

**Comment by EvanBoyle** [comment]
I'll work on the nodejs portion of this today.

**Comment by komalali** [comment]
@justinvp the code snippet from PEP 549 "works" in that the pulumi program does run and give the correct result. However, we lose autocomplete in both pycharm and vscode. And VSCode actually straight up says it's "wrong".

<img width="1071" alt="Screen Shot 2021-06-24 at 1 28 40 PM" src="https://user-images.githubusercontent.com/24326455/123316763-d83d4400-d4f2-11eb-83f0-9b37821a5fb5.png">

Since this has been deprioritized I'll drop it for now and maybe pick it back up eventually.

**Comment by justinvp** [comment]
When we have a chance to get back to this, I wonder if adding a `__init__.pyi` [stub file](https://mypy.readthedocs.io/en/stable/stubs.html) in the `config` directory with definitions that look like regular variables would help the IDEs.

And if that works, wondering if doing something similar would help more generally with the now lazy-loaded submodules that Anton added recently, since the [`if typing.TYPE_CHECKING:`](https://github.com/pulumi/pulumi-aws/blob/cabcec601b3724d916306d1dfe145ace0f7848fe/sdk/python/pulumi_aws/__init__.py#L30) helps mypy, but doesn't help the IDEs.

**Comment by t0yv0** [comment]
IDE discussion now in https://github.com/pulumi/pulumi/issues/7367 (fully orthogonal to the issue here I think).

Now, lazy module loading interaction with global state is pretty unfortunate. If there's at all a chance to remove global state like the issue description is indicating, I'd be a huge fan of trying to do that.

**Comment by komalali** [comment]
@justinvp So, the stub file works in both IDEs 🎉. 

There is a slight complication that we have lost config/outputs since we are now overwriting the entire config module. I'm trying to figure out the user-impact of not having autocomplete for output types... IIRC output types aren't really used directly by users so that should be fine right?

**Comment by justinvp** [comment]
> There is a slight complication that we have lost config/outputs

@komalali, did you find a way to address this? Does it mean we'd have to also have a stub file for `config/outputs`? Importing `outputs` from `__init__.pyi` doesn't "just work"?

**Comment by komalali** [comment]
>did you find a way to address this?

It seems I was wrong about this being an issue... it appears to "just work" as-is, no need to import outputs into `__init__.pyi`, just into `__init__.py` for config.

**Comment by justinvp** [comment]
.NET already uses properties for these, it’s just currently initializing the value up-front once, so we should be able to make these dynamic in a fully compatible way. Though, it does have setters, so to be fully compatible, we will have to allow users to still set a value.

Here's basically what it looks like for [AWS](https://github.com/pulumi/pulumi-aws/blob/master/sdk/dotnet/Config/Config.cs) (a few variables):

```c#
namespace Pulumi.Aws
{
    public static class Config
    {
        private static readonly Pulumi.Config __config = new Pulumi.Config("aws");

        public static string? AccessKey { get; set; } = __config.Get("accessKey");

        public static bool? S3ForcePathStyle { get; set; } = __config.GetBoolean("s3ForcePathStyle");

        public static bool? SkipCredentialsValidation { get; set; } = __config.GetBoolean("skipCredentialsValidation") ?? true;
    }
}
```

We could change it to look something like this:

```c#
namespace Pulumi.Aws
{
    public static class Config
    {
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "IDE1006", Justification =
            "Double underscore prefix used to avoid conflicts with variable names.")]
        private sealed class __Value<T>
        {
            private readonly Func<T> _getter;
            private T _value = default!;
            private bool _set;

            public __Value(Func<T> getter)
            {
                _getter = getter;
            }

            public T Get() => _set ? _value : _getter();

            public void Set(T value)
            {
                _value = value;
                _set = true;
            }
        }

        private static readonly Pulumi.Config __config = new Pulumi.Config("aws");

        private static readonly __Value<string?> _accessKey = new __Value<string?>(() => __config.Get("accessKey"));
        public static string? AccessKey
        {
            get => _accessKey.Get();
            set => _accessKey.Set(value);
        }

        private static readonly __Value<bool?> _s3ForcePathStyle = new __Value<bool?>(() => __config.GetBoolean("s3ForcePathStyle"));
        public static bool? S3ForcePathStyle
        {
            get => _s3ForcePathStyle.Get();
            set => _s3ForcePathStyle.Set(value);
        }

        private static readonly __Value<bool?> _skipCredentialsValidation = new __Value<bool?>(() => __config.GetBoolean("skipCredentialsValidation") ?? true);
        public static bool? SkipCredentialsValidation
        {
            get => _skipCredentialsValidation.Get();
            set => _skipCredentialsValidation.Set(value);
        }
    }
}
```

This emits a private nested helper class named `__Value`. This isn't strictly necessary, but keeps the implementation a bit more tidy.

Each property getter will lookup the live value from `__config` by default, unless a value has been explicitly set, then the getter will always return the set value.
